Visco-Plastic material model

naliboff opened the pull request:
A material model combining viscous (dislocation, diffusion or composite creep) and plastic (Drucker Prager) deformation.  This model will be the site for future implementations of peierls creep and brittle/viscous strain weakening.

tjhei commented on source/material_model/visco_plastic.cc:
```diff
@@ -0,0 +1,620 @@
+                                    cohesions[j] * std::cos(phi) + std::max(pressure,0.0) * std::sin(phi) );
+
+
+          // If the viscous stress is great than the yield strength, rescale the viscosity back to yield surface
```
typo, "greater"

tjhei commented on source/material_model/visco_plastic.cc:
```diff
@@ -0,0 +1,620 @@
+    {
+      for (unsigned int i=0; i < in.temperature.size(); ++i)
+        {
+          // const Point<dim> position = in.position[i];
```
remove if not needed

tjhei commented on source/material_model/visco_plastic.cc:
```diff
@@ -0,0 +1,620 @@
+          // const Point<dim> position = in.position[i];
+          const double temperature = in.temperature[i];
+          const double pressure= in.pressure[i];
+          const std::vector<double> composition = in.composition[i];
```
do you need to make a copy here?

naliboff replied:
A copy of the variables for temperature, pressure, and composition?  Yes, it could have be done without, by I do like the simplicity syntax when calling functions below.  I derived this portion of the code from @gassmoeller diffusion_dislocation material model and I think it would be helpful to keep such related portions of the code consistent for new users who want to develop there own material models.

tjhei replied:
The doubles are fine, but I would avoid making copies of ``std::vector`` if you can. You could just change this to a const ref, which behaves identically, except you don't pay for a copy:
``const std::vector<double> & composition = in.composition[i];``

naliboff replied:
Great, I'll add this change in.

tjhei commented on tests/visco_plastic/screen-ouput:
```diff
@@ -0,0 +1,66 @@
+-----------------------------------------------------------------------------
+
+Number of active cells: 10,000 (on 1 levels)
+Number of degrees of freedom: 171,805 (80,802+10,201+40,401+40,401)
```
do you need a problem with that many unknowns? The cheaper the tests are the better.

naliboff replied:
Sure thing, no reason not to drastically reduce the number of active cells.  The update version will have 100 active cells.

tjhei commented on tests/visco_plastic.prm:
```diff
@@ -0,0 +1,202 @@
+end
+
+# Stop at specified time step
+subsection Termination criteria
```
you don't need that because you already set end time to zero.

tjhei commented on source/material_model/visco_plastic.cc:
```diff
@@ -0,0 +1,620 @@
+                                   " geodnynamic simulations.  See, for example, Billen and Hirth "
+                                   " (2007), G3, 8, Q08012."
+                                   " \n\n "
+                                   " One may select to use the diffusion (\\v_{diff}; $n$=1, m!=0), "
```
some latex markers are missing here

tjhei commented:
/run-tests

tjhei commented on include/aspect/material_model/visco_plastic.h:
```diff
@@ -0,0 +1,198 @@
+/*
+  Copyright (C) 2011 - 2015 by the authors of the ASPECT code.
```
2016

tjhei commented on include/aspect/material_model/visco_plastic.h:
```diff
@@ -0,0 +1,198 @@
+     *               \exp\left(\frac{E + PV}{nRT}\right)\ $
+     * where
+     *   where $A$ is the prefactor, $n$ is the stress exponent,
+     *   \dot{\varepsilon}_{ii} is the square root of the deviatoric
```
missing latex formula markers

tjhei commented on include/aspect/material_model/visco_plastic.h:
```diff
@@ -0,0 +1,198 @@
+     * dislocation and diffusion creep:
+     *   $ v = 0.5 * A^{-\frac{1}{n}} * d^{-\frac{m}{n}} *
+     *               \dot{\varepsilon}_{ii}^{\frac{1-n}{n}} *
+     *               \exp\left(\frac{E + PV}{nRT}\right)\ $
```
Remove the backslash at the end. Also consider a latex formula that is on a separate line.

tjhei commented on source/material_model/visco_plastic.cc:
```diff
@@ -0,0 +1,620 @@
+
+
+          // If the viscous stress is great than the yield strength, rescale the viscosity back to yield surface
+          double viscosity_yield = 0.0;
```
skip the assignment to 0.

tjhei commented on source/material_model/visco_plastic.cc:
```diff
@@ -0,0 +1,620 @@
+            {
+              //not strictly correct if thermal expansivities are different, since we are interpreting
+              //these compositions as volume fractions, but the error introduced should not be too bad.
+              const double temperature_factor= (1.0 - thermal_expansivities[j] * (temperature - reference_T));
```
space before =

tjhei commented on source/material_model/visco_plastic.cc:
```diff
@@ -0,0 +1,620 @@
+                             Patterns::List(Patterns::Double(0)),
+                             "List of angles of internal friction, \\phi, for background material and compositional fields, "
+                             "for a total of N+1 values, where N is the number of compositional fields. "
+                             "For a value of zero, in 2D the von Mises criterion is retrieved "
```
comma at the end?

tjhei commented on source/material_model/visco_plastic.cc:
```diff
@@ -0,0 +1,620 @@
+                             Patterns::List(Patterns::Double(0)),
+                             "List of cohesions, $C$, for background material and compositional fields, "
+                             "for a total of N+1 values, where N is the number of compositional fields. "
+                             "The extremely large default cohesion value (1e20 Pa) prevents the viscous stress from"
```
missing space at the end

tjhei commented on source/material_model/visco_plastic.cc:
```diff
@@ -0,0 +1,620 @@
+        prm.enter_subsection ("Visco Plastic");
+        {
+          // Initialise empty vector for compositional field variables
+          std::vector<double> x_values;
```
what is this used for?

naliboff replied:
It is unused, now removed.

tjhei commented on source/material_model/visco_plastic.cc:
```diff
@@ -0,0 +1,620 @@
+                                   " viscous flow laws.  Plasticity limits viscous stresses through "
+                                   " a Drucker Prager yield criterion. The model is incompressible. "
+                                   " Note that this material model is based heavily on the "
+                                   " diffusion_dislocation (Rene Gassmoeller) and drucker_prager "
```
can you use the class name or the text name of the material models instead?

tjhei commented on source/material_model/visco_plastic.cc:
```diff
@@ -0,0 +1,620 @@
+  {
+    ASPECT_REGISTER_MATERIAL_MODEL(ViscoPlastic,
+                                   "visco plastic",
+                                   " An implementation of a visco-plastic rheology with options for "
```
Can you please add the space only at the end of each line?

tjhei commented on source/material_model/visco_plastic.cc:
```diff
@@ -0,0 +1,620 @@
+                                   " \n\n "
+                                   " The value for the components of this formula and additional "
+                                   " parameters are read from the parameter file in subsection "
+                                   " 'Material model/ViscoPlastic'.")
```
that is not correct (missing space).

tjhei commented:
Very nice contribution so far. You can probably also place a copy into cookbooks/future if you don't want to work on a cookbook section right now.

naliboff commented:
@tjhei - thanks for the review and glad it was in reasonably good shape to start with!  I've incorporated all of the comments above, did a new commit and pushed to the changes to my github account.  I think this should update the pull request?

jdannberg commented on source/material_model/visco_plastic.cc:
```diff
@@ -0,0 +1,617 @@
+                                         std::pow(edot_ii,((1. - stress_exponents_dislocation[j])/stress_exponents_dislocation[j]));
+
+          // Composite viscosity
+          double viscosity_composite = (viscosity_diffusion * viscosity_dislocation)/(viscosity_diffusion + viscosity_dislocation);
```
I have a question about your composite diffusion/dislocation rheology:
In the `diffusion_dislocation` material model, we use the dislocation strain rate here instead of `edot_ii` to compute the dislocation viscosity. (v_{eff}=\sigma_{ii}/(2 edot_{ii}), but v_{dis}=\sigma_{ii}/(2 edot_{ii,dis}).) Because we do not have the dislocation strain rate, we have to do some iterations until we haver determined how much of the strain rate is due to diffusion/dislocation creep (the while loop in the `diffusion_dislocation` material model). 
I am not really sure what people normally do, but is there any reason you have done it like this? I guess the additional plasticity would make iterating out which process accommodates the strain even more complicated. 
I was just thinking, if we have two material models that use composite diffusion/dislocation creep, but the implementations are different, we should point that out somewhere and say why they are different. Also, I guess it would be interesting to see how much of a difference it makes. Have you compared your model (without plasticity) to the `diffusion_dislocation` material model?

naliboff replied:
Thanks for bringing this up, how to handle the dislocation/composite rheology is tricky.  My understanding, which certainly could be wrong, is that most codes follow the approach I took and assume the dislocation strain-rate is the same as the background strain-rate (edot_{ii}).  In regions with high-strain rates, this usually translates to dislocation creep having a much lower viscosity than diffusion creep and the composite viscosity is effectively the same as the dislocation creep viscosity.  The weakest mechanism dominates.  

However, I completely agree this is not necessarily a good assumption and the iterations in `diffusion_dislocation` seem like a reasonable way to approximate the relative contributions.  I often only use a dislocation creep rheology for the lithosphere and need to think about composite rheologies in more detail.

I think a good solution would be to add in an option for determining the relative strain-rate between diffusion/dislocation creep (following method in `diffusion_dislocation`) in the case where a composite rheology is selected.  I'll do some testing first to make sure no odd issues  arise when doing this type of iterations in combination with plastic deformation, but I don't anticipate any.  

Does this sound like a good approach?   I'm all for having as many options as possible with this material model.  It can be really informative to see how a small change to the non-linear rheology can affect short- and long-term dynamics.

Thanks again for bringing this up and reviewing the code!

jdannberg replied:
Yes, thanks, that sounds like a great idea!

jdannberg commented on source/material_model/visco_plastic.cc:
```diff
@@ -0,0 +1,617 @@
+                                   "viscous flow laws.  Plasticity limits viscous stresses through "
+                                   "a Drucker Prager yield criterion. The model is incompressible. "
+                                   "Note that this material model is based heavily on the "
+                                   "DiffusionDislocation (Rene Gassmoeller) and DruckerPrager "
```
Also, I believe Bob Myhill (@bobmyhill) wrote the diffusion dislocation material model originally, and Anne Glerum (@anne-glerum) contributed the drucker prager model.

naliboff replied:
Thanks, I will update the names!

jdannberg commented:
@naliboff, thanks for this contribution! Can you also run astyle (using the script `./doc/indent` from the top-level directory of ASPECT) so that all of your files have the indentation we normally use in Aspect?

naliboff commented:
@jdannberg, I just the indent script, will push the modified files up.

gassmoeller commented:
It seems the test output is not correct. @naliboff you could check if the test succeeds on your machine? Before merging we would also need an entry in doc/modules/changes.h (just copy one of the existing one and summarize your new material model in one sentence).

naliboff commented:
@gassmoeller - Thanks! I think the issue is just an improperly named file (screen-output), but I also removed the statistics and depth-average plots and modified/reran the parameter file accordingly.  Test succeeded on my machine.  Pushing changes now.

Question regarding making the test output .. I ran the model and then copied the parameter file and log.txt (renamed to screen-output) to the tests folder.  Is this the correct procedure?

naliboff commented:
@gassmoeller - seems like a test is still failing? Could this be a dealii version issue?  I had been using deal.ii 8.3.0, but am now recompiling with 8.4.1.  Will push the changes up as soon as I re-run the test.

tjhei commented:
> @gassmoeller - seems like a test is still failing? Could this be a dealii version issue? I had been using deal.ii 8.3.0, but am now recompiling with 8.4.1. Will push the changes up as soon as I re-run the test.

Many tests are sensitive to the OS/compiler/library versions/deal.II versions. This is unfortunate because it means it is hard to generate the "correct" output that the tester will accept. But you can get a changes.diff file from the tester with the correct output that will make the test pass. So just click on "details", download the diff and apply with ``patch -p1 <changes.diff``.

In your case this still won't work though, because you caught us in a weird transition state from deal.II 8.3 to 8.4. You first need to rebase to the current ASPECT, wait for the tester to rerun, and then copy the test output. Let us know if you need help on how to do that.

naliboff commented:
@tjhei - gotcha, rebased and am re-running tests on my end. Question regarding copying the test output: do you mean the output generated by the tester (my machine or github tester) or simply re-run the test as a normal model with the rebased version and copy the output to 'screen-output' file in my local test folder?

naliboff commented:
@tjhei - well, I rebased my branch to upstream master (git rebase upstream/master), ran all tests on the local machine (all passed), updated 'screen-output' based on re-running the test model and pushed the commits.  

Was this not the right procedure given the pull request now has everything from the master version via the rebase (i.e. the pull request looks like a mess)?  I guess the right procedure would to rerun the test in a different, but rebased branch and then copy the test results into this branch?  Sorry for the confusion, I can make a pull request if needed.

tjhei commented:
> updated 'screen-output' based on re-running the test model and pushed the commits.

No, you need to take the "official" output from the tester by downloading and applying the diff.


> (i.e. the pull request looks like a mess)

Yeah, this is a pretty amazing mess :-) You managed to have the commits twice in your history.

It looks like if you do
```
git branch backup_current_state
git reset --hard 30e480e6b1
git remote update
git rebase upstream/master
git push -f mygithub HEAD
```
should give you something like https://github.com/tjhei/aspect/commits/pr865

Then wait for the tester, apply the patch, and push again.

naliboff commented:
Applied the fix, thank you!  Will push once I've applied the test output patch.

naliboff commented:
@tjhei - aspect-8.4.1 test is passing, 8.4.1b is failing but the error is associated with gccpetsc and the free_surface_adaptive_blob_petsc test.  Should I still apply these "diffs" to the appropriate tests (free_surface_adaptive_blob_petsc) or are things good to go on my end?

tjhei commented:
> 8.4.1b is failing

No, that is my fault (I set that thing up yesterday).

tjhei commented:
The changelog is missed up (please remove everything except you one new entry), can you fix that please?

naliboff commented:
I cleaned up the changelog entry: now only contributions from v1.4 (i.e. what was in changes.h in master) and my entry are present.

tjhei commented:
:+1:, sorry for the complications caused by changes in tester/deal.II versions, etc..

naliboff commented:
Great, thanks for all the help!  Sorry for the complications caused by haphazard rebase.
